### PR TITLE
`network.informo.article` matrix event

### PR DESCRIPTION
**Pull request by GordonF42**
This defines the content of an Informo article event

Currently this implies having two "formats" for `network.informo.article` events

- Signed article:
```json
{
    "content": {
        "algorithm": "ed25519"
        "sender_key": "IlRMeOPX2e0MurIyfWEucYBRVOEEUMrOHqn/8mLqMjA"
        "signature": "0a1df56f1c3ab5b1"
        "signed": {
            "title": "Lorem ipsum",
            "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."
        }
    }
}
```

- Unsigned article:
```json
{
    "content": {
        "title": "Lorem ipsum",
        "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."
    }
}
```

To know if an article event is signed or not, the clients will need to check the presence if the `signed` key. This may not be perfect but it is quite handy since the signed event on Informo follows a common structure that can be easily decapsulated.

Let me know what you think ;)

### CODE REVIEW

**Comment by vabd** [comment]
Didn't we previously agree on the fact that the events would be of the `network.informo.article.SOURCE_SLUG`? That way, it makes it easier to filter the request to Matrix's `/sync` and `/messages` endpoints based on the user's subscriptions in the client.

**Comment by GordonF42** [comment]
This was the POC's behaviour, but changed it because of the following reasons:
- [filters allows to filter by event sender](https://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-r0-user-userid-filter), the same way we can filter event types.
- It is forbidden to publish articles in the name of another source. With the previous behaviour any user could send `network.informo.article.ACME` and client had to check whether the user was ACME or not. 
- It strengthen the relation between the source and its articles: If a source user sends an article event, it's the source's article and no one else's. There's no logical way around that requirement.

**Comment by GordonF42** [comment]
One thing bothers me:
We discussed previously about setting the article's locale in the source configuration, i.e. one source always send articles in one single language, and if an entity wants to provide multilingual content they'll need to set-up multiple sources.

I think this will quickly become a mess when entities will need to create many sources with different names, like _AcmeNews English_, _AcmeNews French_,  _AcmeNews German_, ... and currently we don't provide any easy way to know which locales are provided by ACMENews (except by judging source names). It also makes phishing a lot easier since anybody can register _AcmeNews Spanish_ and it would look quite legit.

The phishing issue can be resolved by requiring ACMENews to act as a TA trusting its sources. I think this is patchy since a TA's original purpose is to build trust between separated organisations and not inside one single organisation.

---

I'd rather go for a one source and multiple localized articles approach:

- The source declares a list of provided locales, and send all its localized articles with the same user
- Article events contains a `locale` field.
- When the user wants to subscribe to the source, they choose one or more locales, and the client hides articles in an undesired locales.

File: content/information-distribution/article.md [hunk 1]
```diff
@@ -30,4 +30,48 @@ If the original news item contains media, these media **should** be uploaded to
 the node using [Matrix's content repository
 module](https://matrix.org/docs/spec/client_server/r0.4.0.html#id112).
 
-2️⃣: rest of the page
+## Matrix event `network.informo.article`
+
+This message event is meant to be sent by a source Matrix user. If the article
+is signed, it **must** be wrapped into a [Signed Matrix
```
Comment (vabd) [inline content/information-distribution/article.md]: "signed" shouldn't be capitalised.

File: content/information-distribution/article.md [hunk 2]
```diff
@@ -30,4 +30,48 @@ If the original news item contains media, these media **should** be uploaded to
...
+source registers itself afterwards, its previously sent articles **should**
+become visible.
+
+
+### Event data
+
+|      Parameter      |   Type   | Req. |                                                                            Description                                                                            |
+| ------------------- | -------- | :--: | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `title`             | `string` |  x   | Article's headline                                                                                                                                                |
+| `href`              | `string` |      | Link to the article's original post (in case the article was sent to informo from an existing website)                                                            |
```
Comment (vabd) [inline content/information-distribution/article.md]: "Informo"'s "I" should be capitalised.

File: content/information-distribution/article.md [hunk 3]
```diff
@@ -30,4 +30,48 @@ If the original news item contains media, these media **should** be uploaded to
...
+        "algorithm": "ed25519"
+        "sender_key": "IlRMeOPX2e0MurIyfWEucYBRVOEEUMrOHqn/8mLqMjA"
+        "signature": "0a1df56f1c3ab5b1"
+        "signed": {
+            "title": "Lorem ipsum",
+            "content": "Lorem ipsum dolor sit amet, consectetur adipiscing elit."
+        }
+    }
+}
+```
```
Comment (vabd) [inline content/information-distribution/article.md]: JSON isn't valid (lines don't end with `,`.

File: content/information-distribution/article.md [hunk 4]
```diff
@@ -30,4 +30,48 @@ If the original news item contains media, these media **should** be uploaded to
...
+| `title`             | `string` |  x   | Article's headline                                                                                                                                                |
+| `href`              | `string` |      | Link to the article's original post (in case the article was sent to informo from an existing website)                                                            |
+| `short_description` | `string` |      | Short description or introduction to the article.                                                                                                                 |
+| `author`            | `string` |      | Full name of the article's author (when a single Informo source aggregates multiple writers).                                                                     |
+| `thumbnail`         | `string` |      | Preview image for the article. Must be a `mxc://` url.                                                                                                            |
```
Comment (vabd) [inline content/information-distribution/article.md]: Might be worth adding a link to the mxc specs (which would be https://matrix.org/docs/spec/client_server/r0.4.0.html#id112)

File: content/information-distribution/article.md [hunk 5]
```diff
@@ -30,4 +30,48 @@ If the original news item contains media, these media **should** be uploaded to
...
+| `date`              | `int`    |      | Timestamp in milliseconds when the article was published. If not provided clients should fall-back to the timestamp when the matrix event was sent.               |
```
Comment (vabd) [inline content/information-distribution/article.md]: Not sure of the phrasing here. I'd phrase it like this:

```suggestion
| `date`              | `int`    |      | Timestamp in milliseconds of the article's publication. If not provided clients should fall back to the Matrix event's creation timestamp.               |
```

**Review by vabd — CHANGES_REQUESTED** [review]
Otherwise LGTM.

**Comment by vabd** [comment]
> I'd rather go for a one source and multiple localized articles approach:
> 
> * The source declares a list of provided locales, and send all its localized articles with the same user
> 
> * Article events contains a `locale` field.
> 
> * When the user wants to subscribe to the source, they choose one or more locales, and the client hides articles in an undesired locales.

My concern here is that Informo is likely to be used in situations with bandwidth is very limited. This, plus considering that the use case of people wanting to read articles from a single source in more than one language would be pretty rare, and more than one almost non-existent, I really see this handling of l10n creating huge payloads sent to the user, with more than 50% of it being useless to the user it's being sent to most of the time, in situations where we want to keep payloads at minimal sizes.

On top of that, most websites available in more than one language have different websites/RSS feeds/editorial teams/etc. for each language, and centralising all of them on a single input might make add some useless complexity to the technical implementation.

One alternative way to handle l10n (which just occurred to me), and fixes both the issue of centralising localised websites of a single media and the one of providing payload of decent sizes would be to spec a new type of entity, which would be sort of "super sources" (which would not be the final name at all). This source would be the one trusted by TAs, and would declare a list of Matrix users handling the publication of every localised website. The idea would then be to place these localised sources at the same level in trust chains as their super source, since both are managed by the same organisation (or individual), in order for client implementations to make the whole thing transparent to users. Localised sources would have to emit a registration event including a signature from their super source rather than a TA. Of course, a super source could also reference itself as the publisher for a given locale (which would ease the migration for a news source opening a new website in a different language).

What would you think about that?

**Comment by GordonF42** [comment]
Your super source / sub source idea does solve the problem, and I think it needs more in-depth investigation in a separate PR.

For now I think we should keep the current data model and extend it afterwards when the source data model is more mature.

**Comment by vabd** [comment]
> I think it needs more in-depth investigation in a separate PR.
> 
> For now I think we should keep the current data model and extend it afterwards when the source data model is more mature.

Agreed (which is why I mentioned it in a comment rather than as part of the review).

FTR I'll start drafting a SCS for this today.

File: content/information-distribution/article.md (hunk 1 above)
Comment (vabd) [reply content/information-distribution/article.md]: This part should be left as is until #8 gets merged. Afterwards, two issues need to be addressed:

1. "If the article is signed" => articles are always signed
2. No mention who signs the event

Both can be solved by removing the paragraph, or rephrasing it, since, as it currently is phrased, #8 makes it redundant.

File: content/information-distribution/article.md (hunk 2 above)
Comment (vabd) [reply content/information-distribution/article.md]: Also, `s/Link to/URL of/`

File: content/information-distribution/article.md (hunk 4 above)
Comment (vabd) [reply content/information-distribution/article.md]: Also, please write "URL" in upper case since it's an acronym.

File: content/information-distribution/article.md (hunk 1 above)
Comment (GordonF42) [reply content/information-distribution/article.md]: IMO we should consider #8 merged as it will likely be merged before this PR (thus the link to the signing page)

Comment (vabd) [reply content/information-distribution/article.md]: Sure, I was suggesting to do it that way (then merge master into this branch) so we can have a full picture of the page's final state.

File: content/information-distribution/article.md [hunk 6]
```diff
@@ -32,9 +32,9 @@ module](https://matrix.org/docs/spec/client_server/r0.4.0.html#id112).
 
 ## Matrix event `network.informo.article`
 
-This message event is meant to be sent by a source Matrix user. If the article
-is signed, it **must** be wrapped into a [Signed Matrix
-event](/information-distribution/signature/#signed-matrix-event).
+This message event **must** be sent and signed by a source Matrix user. See
+[signed Matrix event](/information-distribution/signature/#signed-matrix-event)
```
Comment (vabd) [inline content/information-distribution/article.md:27]: Sorry to be picky here, but I think the rendered version of this page might not make the face that we refer to another page obvious enough. I'd rephrase that into "See the [signed Matrix event] page for more information...".

File: content/information-distribution/article.md [hunk 7]
```diff
@@ -43,16 +43,16 @@ become visible.
...
+| `title`             | `string` |  x   | Article's headline.                                                                                                                                 |
+| `href`              | `string` |      | URL of the article's original post (in case the article was sent to Informo from an existing website).                                              |
+| `short_description` | `string` |      | Short description or introduction to the article.                                                                                                   |
+| `author`            | `string` |      | Full name of the article's author (when a single Informo source aggregates multiple writers).                                                       |
+| `thumbnail`         | `string` |      | Preview image for the article. Must be a [`mxc://` url](https://matrix.org/docs/spec/client_server/r0.4.0.html#id112).                                                                                              |
```
Comment (vabd) [inline content/information-distribution/article.md]: To forgot to spell URL in upper case here.

File: content/information-distribution/article.md [hunk 8]
```diff
@@ -43,16 +43,16 @@ become visible.
...
+| `date`              | `int`    |      | Timestamp in milliseconds when the article was published. If not provided clients should fall-back to the timestamp when the matrix event was sent. |
```
Comment (vabd) [inline content/information-distribution/article.md]: Same comment as from my previous review.

**Review by vabd — CHANGES_REQUESTED** [review]
Me being picky, lgtm otherwise.

**Review by vabd — APPROVED** [review]

**Comment by vabd** [comment]
lgtm, merging as it's already past the 14 day review period.
